adagrams/game.py

```diff
@@ -1,11 +1,64 @@
+from random import sample
 def draw_letters():
-    pass
+    hand = sample(['A','B','C','D','E','F','G','H','I','J','K','L',\
+                   'M','N','O','P','Q','R','S','T','U','V','W','X',\
+                   'Y','Z'], counts = [9,2,2,4,12,2,3,2,9,1,1,4,2,6,\
+                                       8,2,1,6,4,6,4,2,2,1,2,1], k = 10)
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letter_bank_dict = {}
+    for letter in letter_bank:
+        if letter not in letter_bank_dict:
+            letter_bank_dict[letter] = 1
+        else:
+            letter_bank_dict[letter] += 1
+    for letter in word.upper():
+        if letter not in letter_bank_dict:
+            return False
+        elif letter_bank_dict[letter] == 0:
+            return False
+        else:
+            letter_bank_dict[letter] -= 1
+    return True
+
 
 def score_word(word):
-    pass
+    scores = {
+        "AEIOULNRST" : 1,
+        "DG" : 2,
+        "BCMP" : 3,
+        "FHVWY" : 4,
+        "K" : 5,
+        "JX" : 8,
+        "QZ" : 10
+    }
+
+    score = [val for letter in word.upper() for key, val in scores.items() if letter in key]
```

Comment:
I'll point out that this comment about space complexity does _not_ apply if rather than a comprehension, you use a generator (working from my improved "flat" dictionary implementation):

```python
score = sum(scores[letter] for letter in word.upper())
```

Note that rather than having square brackets, I simply put the same expression into a call to sum. When it is like this, rather than a list comprehension it becomes a "generator". Which is essentially equivalent to this function:

```python
def score_letters(scores, word):
    for letter in word.upper():
        yield scores[letter]
```

And then you can imagine it as if `sum` was written like this:

```python
def sum(generator):
    total_sum = 0
    for value in generator:
        total_sum += value
    return sum
```

So basically my original use of the generator would become `score = sum(score_letters(scores, word))`, and this ends up doing the exact same thing as in my suggestion, just divided across two functions.

Here's a nice article on generators, because they are a very useful part of Python (and other languages have similar features): https://realpython.com/introduction-to-python-generators/

Reply:
In fact, you can think of a comprehension as just a shorthand way of actually allocating the results of a generator expression into a data structure!